### PR TITLE
Implement fast bazel target lookup with index caching

### PR DESCRIPTION
**Pull request by titusfortner**
### **User description**
I've tried several times to improve how we determine which targets to run tests against.

The options are all either inaccurate (just run everything) or slow (so slow they were timing out).
This is the hybrid approach. Do the slow thing 2x a day when scheduled CI runs, and the fast thing every other time. Should be good enough.

The pre-release PR found [2437 matching targets](https://github.com/SeleniumHQ/selenium/actions/runs/21107149612/job/60700310765), this PR finds the correct 570.

### 💥 What does this PR do?

* Adds Rake task `bazel:build_test_index` to index test targets based on affected BUILD packages
```
./go bazel:build_test_index            # --> path: 'build/bazel-test-target-index.json'
./go bazel:build_test_index index.json # --> path: 'index.json'
```
* Adds Rake task `bazel:affected_targets` to provide a list of test targets that apply
```
./go bazel:affected_targets            # --> HEAD^..HEAD (last commit)
./go bazel:affected_targets abc123     # --> abc123^..abc123 (that commit)
./go bazel:affected_targets abc..def   # --> explicit range
```

* When ci.yml event is scheduled it runs the indexing
* When ci.yml event is anything else it checks the index for matching targets 
* Indexes get zipped and stored in GitHub cache (very small, very quick to zip/unzip)
* Deleted scripts/github-actions/check-bazel-targets.sh (replaced by Rake tasks)
* Added cache-name input to bazel.yml workflow
* Simplified per-binding conditionals in ci.yml

Edge Cases:
1. Add src file not in index: `#find_bazel_package` locates the package and runs tests for that package
2. Delete src file: `#find_bazel_package` locates the package and runs tests for that package
3. Add test file not in index: all changed test files get corresponding targets added directly
4. Delete test file: ignored
5. Add new package: generates target list from package (slow but infrequent)
6. .NET tests depend on Java server causing all .NET tests to run each time java server code is changed; since .NET does not have any tests running against a grid, this does a complete filter
7. If one test file is updated only one test file should run

### 🔧 Implementation Notes

I've spent way too long on this. I tried the tools, I tried various shortcuts, no good way to get accuracy with any speed.
I went back and forth on shell vs python vs rake, and Rake gives me the most flexibility, and lets us easily do this from local or on CI

### 💡 Additional Considerations

JavaScript will need to be added later
Rust is ignored in the indexing because we aren't toggling rust tests from this
This doesn't actually use the targets in the ci.yml yet, so we're still just seeing if there are any and running everything.
Actually limiting things is next

### 🔄 Types of changes

- New feature (non-breaking change which adds functionality *and tests!*)

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Replace slow shell script with fast Rake-based Bazel target lookup using index caching

- Add `bazel:build_test_index` task to generate test target index from BUILD packages

- Add `bazel:affected_targets` task to find affected targets using index or fallback

- Simplify CI conditionals by splitting target output into per-binding flags

- Switch index generation from macOS to Linux for faster, cheaper builds

- Support workflow dispatch to rebuild index and run all tests

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Changed Files"] --> B["affected_targets Task"]
  B --> C{"Index Exists?"}
  C -->|Yes| D["Query Index"]
  C -->|No| E["Directory Fallback"]
  D --> F["Test Targets"]
  E --> F
  G["build_test_index Task"] --> H["Generate Index from BUILD Packages"]
  H --> I["Cache Index"]
  I --> D
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>check-bazel-targets.sh</strong><dd><code>Remove slow shell-based target lookup script</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

scripts/github-actions/check-bazel-targets.sh

<ul><li>Deleted entire shell script that performed slow Bazel target lookup<br> <li> Functionality replaced by Rake tasks in Rakefile<br> <li> Script used rdeps queries which were timing out</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16974/files#diff-6b8abebbae3ba9df24a8d73fd37c3b4e4b6787e8b5fc671e2507bec1e492e975">+0/-108</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>bazel.yml</strong><dd><code>Add cache restore support for index files</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/bazel.yml

<ul><li>Added <code>cache-name</code> input parameter for cache restore functionality<br> <li> Added cache restore step using actions/cache/restore@v4<br> <li> Allows workflows to restore pre-built index files</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16974/files#diff-85e6e6bfc3047eb4fe69bf0f06238ddd442edf74266ad3a964ebfba26f46923c">+12/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ci-build-index.yml</strong><dd><code>Switch index generation to Linux platform</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/ci-build-index.yml

<ul><li>Changed index build OS from macOS to Linux<br> <li> Improves build speed and reduces CI costs<br> <li> Requires trunk fix for macOS-only archive rules</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16974/files#diff-af6fa0c76fca0cbf259f6438789f3347aa92a11ed8c559d955a129bccc3df010">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ci.yml</strong><dd><code>Refactor CI workflow to use Rake tasks and per-binding flags</code></dd></summary>
<hr>

.github/workflows/ci.yml

<ul><li>Updated build-index job to run on schedule or workflow_dispatch events<br> <li> Replaced shell script invocation with inline Rake task calls<br> <li> Refactored read-targets job to output per-binding flags instead of <br>combined targets<br> <li> Simplified per-binding conditionals from complex multi-line <br>expressions to simple flag checks<br> <li> Added cache-name parameter to check job for index restoration</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16974/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f">+41/-48</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Rakefile</strong><dd><code>Add Rake tasks for fast Bazel target lookup with index</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Rakefile

<ul><li>Commented out lint task implementation (formatting, mypy, ruff, steep <br>checks)<br> <li> Added <code>bazel:affected_targets</code> task to find test targets affected by <br>file changes<br> <li> Added <code>bazel:build_test_index</code> task to generate index from BUILD <br>packages<br> <li> Added helper functions: <code>find_bazel_package</code>, <br><code>affected_targets_with_index</code>, <code>targets_from_lookup</code>, <code>query_package_dep</code>, <br><code>targets_from_tests</code>, <code>affected_targets_fallback</code><br> <li> Supports index-based lookup with directory fallback for missing <br>packages<br> <li> Handles edge cases like test files, library files, and Java/dotnet <br>dependencies</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16974/files#diff-ee98e028c59b193d58fde56ab4daf54d43c486ae674e63d50ddf300b07943e0f">+144/-17</a></td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+  task :build_test_index, [:index_file] do |_task, args|
+    output = args[:index_file] || 'build/bazel-test-target-index.json'
+
+    index = {}
+    tests = []
+
+    exclude_tags = %w[manual spotbugs ie]
+    all_bindings = BINDING_TARGETS.values.join(' + ')
+    tag_exclusions = exclude_tags.map { |tag| "except attr(tags, #{tag}, #{all_bindings})" }.join(' ')
+    kind = '_test' # do not match test_suite or pytest_runner
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: if you just use "test" it matches test_suite and pytest_runner. If you set it to _test it will not match those things.

File: Rakefile [hunk 2]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+  test_targets = index[pkg] || query_package_dep(pkg)
+
+  # dotnet tests depend on java server, but there are no remote tests, so safe to ignore
+  filepath.start_with?('java/') ? test_targets.reject { |t| t.start_with?('//dotnet/') } : test_targets
+end
+
+def query_package_dep(pkg)
+  puts "Package not in index, querying deps: //#{pkg}"
+  targets = []
+  Bazel.execute('query', ['--output=label'], "kind('.*_test', deps(//#{pkg}:all))") do |out|
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: pkg comes from find_bazel_package which returns repo directory paths, not user input

File: Rakefile [hunk 3]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+    return nil if path == '.'
+
+    path = File.dirname(path)
+  end
+  nil
+end
+
+def affected_targets_with_index(changed_files, index_file)
+  puts "Using index: #{index_file}"
+  index = JSON.parse(File.read(index_file))
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: it's fine

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -26,8 +31,19 @@ jobs:
...
+      run: |
+        if [ "${{ github.event_name }}" == "schedule" ] || [ "${{ github.event_name }}" == "workflow_dispatch" ] || [ "${{ github.event_name }}" == "workflow_call" ]; then
+          echo "//java/... //py/... //rb/... //dotnet/... //rust/..." > bazel-targets.txt
+        else
+          if [ -f bazel-test-target-index.gz ]; then
+            mkdir -p build
+            gunzip -c bazel-test-target-index.gz > build/bazel-test-target-index.json
+            echo "Restored cached index"
+          fi
+          ./go bazel:affected_targets "${{ github.event.pull_request.base.sha || github.event.before }}..HEAD"
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply .github/workflows/ci.yml]: Checkout action sets HEAD to the correct ref value so long as it is available, hence fetch-depth

File: Rakefile [hunk 5]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+    index_file = args[:index_file] || 'build/bazel-test-target-index.json'
+    base_rev, head_rev = if range.include?('..')
+                           range.split('..', 2)
+                         else
+                           ["#{range}^", range]
+                         end
+
+    puts "Commit range: #{base_rev}..#{head_rev}"
+
+    changed_files = `git diff --name-only #{base_rev} #{head_rev}`.split("\n").map(&:strip).reject(&:empty?)
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: SHAs are hex-only from GitHub API

File: Rakefile [hunk 6]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+  test_files.select! { |f| File.exist?(f) }
+  return [] if test_files.empty?
+
+  query = test_files.filter_map { |f|
+    pkg = find_bazel_package(f)
+    next if pkg.nil?
+
+    # Bazel srcs often use paths relative to the package, not basenames.
+    rel = f.sub(%r{\A#{Regexp.escape(pkg)}/}, '')
+    "attr(srcs, '#{rel}', //#{pkg}:*)"
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: just don't put quotes in file names

File: rb/.rubocop.yml [hunk 7]
```diff
@@ -160,6 +160,10 @@ Lint/Debugger:
 Lint/UselessConstantScoping:
   Enabled: false
 
+Lint/RedundantRequireStatement:
+  Exclude:
+    - '../Rakefile'
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply rb/.rubocop.yml]: Path is relative to config file location, this works correctly in all contexts

File: Rakefile [hunk 8]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+    targets = out.lines.map(&:strip).select { |l| l.start_with?('//') }
+  end
+  targets
+end
+
+def targets_from_tests(test_files)
+  test_files.select! { |f| File.exist?(f) }
+  return [] if test_files.empty?
+
+  query = test_files.filter_map { |f|
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: we create it then don't use it. I have more tolerance for mutation than others apparently.

File: Rakefile [hunk 9]
```diff
@@ -1640,3 +1641,181 @@ def update_changelog(version, language, path, changelog, header)
...
+}.freeze
+
+namespace :bazel do
+  # ./go bazel:affected_targets           --> HEAD^..HEAD (last commit)
+  # ./go bazel:affected_targets abc123    --> abc123^..abc123 (that commit)
+  # ./go bazel:affected_targets abc..def  --> explicit range
+  desc 'Find test targets affected by changes between revisions'
+  task :affected_targets, %i[commit_range index_file] do |_task, args|
+    range = args[:commit_range] || 'HEAD'
+    index_file = args[:index_file] || 'build/bazel-test-target-index.json'
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: you are hallucinating

File: .github/workflows/bazel.yml [hunk 10]
```diff
@@ -100,6 +105,13 @@ jobs:
         run: git pull origin "$GIT_REF"
         env:
           GIT_REF: ${{ github.ref }}
+      - name: Restore cache
+        if: inputs.cache-name != ''
+        uses: actions/cache/restore@v4
+        with:
+          path: ${{ inputs.cache-name }}.gz
+          key: ${{ inputs.cache-name }}-
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply .github/workflows/bazel.yml:113]: we need key for restore-keys to work, just because it will never match (intended) doesn't mean it is wrong or should be deleted.

**Comment by titusfortner** [comment]
I committed 0850cd084b0b3feec7cdc6199cfd52613f4e60f2 which does the indexing part of this. Once a scheduled CI job has run building the index without issues, I'll merge the rest of this.

**Comment by titusfortner** [comment]
I'm putting in draft until #16985 merges

**Comment by titusfortner** [comment]
It should index on linux now.

File: Rakefile [hunk 11]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
...
+  # dotnet tests depend on java server, but there are no remote tests, so safe to ignore
+  filepath.start_with?('java/') ? test_targets.reject { |t| t.start_with?('//dotnet/') } : test_targets
+end
+
+def query_package_dep(pkg)
+  # Root package is empty string, not '.'
+  pkg = '' if pkg == '.'
+  puts "Package not in index, querying deps: //#{pkg}"
+  targets = []
+  Bazel.execute('query', ['--output=label'], "kind('.*_test', deps(//#{pkg}:all))") do |out|
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: Package names from git/filesystem, not user input

File: Rakefile [hunk 12]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
...
+    path = File.dirname(path)
+  end
+  nil
+end
+
+def affected_targets_with_index(changed_files, index_file)
+  puts "Using index: #{index_file}"
+  begin
+    index = JSON.parse(File.read(index_file))
+  rescue JSON::ParserError => e
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: overkill

File: Rakefile [hunk 13]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
   label = label.sub(%r{^@selenium//}, '').sub(%r{^//}, '')
   label.split(':').first
 end
+
+def find_bazel_package(filepath)
+  path = File.dirname(filepath)
+  until path.empty?
+    return path if File.exist?(File.join(path, 'BUILD.bazel')) || File.exist?(File.join(path, 'BUILD'))
+    return nil if path == '.'
+
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: preference

File: Rakefile [hunk 14]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
...
+  Bazel.execute('query', ['--output=label'], query) do |out|
+    targets = out.lines.map(&:strip).select { |l| l.start_with?('//') }
+  end
+  targets
+end
+
+def affected_targets_fallback(changed_files)
+  targets = Set.new
+  top_level_dirs = changed_files.map { |f| f.split('/').first }.uniq
+
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: it'll change soon enough

File: Rakefile [hunk 15]
```diff
@@ -1357,11 +1357,50 @@ BINDING_TARGETS = {
...
+      targets.sort.each { |t| puts t }
+    end
+  end
+
+  # ./go bazel:build_test_index           --> 'build/bazel-test-target-index'
+  # ./go bazel:build_test_index foo.json  --> 'foo.json'
   desc 'Build test target index for faster affected target lookup'
   task :build_test_index, [:index_file] do |_task, args|
-    output = args[:index_file] || 'build/bazel-test-target-index.json'
+    output = args[:index_file] || 'build/bazel-test-target-index'
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: tbh I left it off in case that was why it was failing in CI. I got it working and don't want to mess with it again.

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -41,83 +55,62 @@ jobs:
...
+            local pattern=$1 tag=$2
+            if [[ "$targets" == *"$pattern"* ]] || [[ "$COMMIT_MESSAGES" == *"[$tag]"* ]] || [[ "$PR_TITLE" == *"[$tag]"* ]]; then
+              echo "$tag=true" >> "$GITHUB_OUTPUT"
+            fi
+          }
+          check_binding "//java/" "java"
+          check_binding "//py:" "py"
+          check_binding "//rb/" "rb"
+          check_binding "//dotnet/" "dotnet"
+          check_binding "//rust/" "rust"
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply .github/workflows/ci.yml:85]: future

File: Rakefile [hunk 17]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
...
+    targets = out.lines.map(&:strip).select { |l| l.start_with?('//') }
+  end
+  targets
+end
+
+def targets_from_tests(test_files)
+  test_files.select! { |f| File.exist?(f) }
+  return [] if test_files.empty?
+
+  query = test_files.filter_map { |f|
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: meh

File: Rakefile [hunk 18]
```diff
@@ -1408,3 +1447,93 @@ def bazel_label_to_package(label)
...
+    pkg = find_bazel_package(f)
+    next if pkg.nil?
+
+    # Bazel srcs often use paths relative to the package, not basenames.
+    rel = f.sub(%r{\A#{Regexp.escape(pkg)}/}, '')
+    "attr(srcs, '#{rel}', //#{pkg}:*)"
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply Rakefile]: aggressive

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -31,8 +31,22 @@ jobs:
...
+        [ "${{ github.event_name }}" == "workflow_call" ] || \
+        [ "${{ github.event_name }}" == "workflow_dispatch" ]; then
+          echo "Running all targets for ${{ github.event_name }} event"
+          echo "//java/... //py/... //rb/... //dotnet/... //rust/..." > bazel-targets.txt
+        else
+          BASE_SHA="${{ github.event.pull_request.base.sha || github.event.before }}"
+          if [ -n "$BASE_SHA" ]; then
+            ./go bazel:affected_targets "${BASE_SHA}..HEAD" bazel-test-target-index
+          else
+            ./go bazel:affected_targets bazel-test-target-index
```
[the remark this answers is not in the document]
Comment (titusfortner) [reply .github/workflows/ci.yml:47]: that's a workflow_dispatch which hits the if. It can't have both multiple and 0.
